Content.Server/Chat/V2/Repository/ChatRepository.cs

```diff
@@ -0,0 +1,238 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Numerics;
+using Content.Shared.Chat.V2;
+using Content.Shared.Chat.V2.Repository;
+using Robust.Server.Player;
+using Robust.Shared.Replays;
+
+namespace Content.Server.Chat.V2.Repository;
+
+/// <summary>
+/// Stores <see cref="IChatEvent"/>, gives them UIDs, and issues <see cref="MessageCreatedEvent"/>.
+/// Allows for deletion of messages.
+/// </summary>
+public sealed class ChatRepository : EntitySystem
+{
+    [Dependency] private readonly IReplayRecordingManager _replay = default!;
+    [Dependency] private readonly IPlayerManager _player = default!;
+
+    // Clocks should start at 1, as 0 indicates "clock not set" or "clock forgotten to be set by bad programmer".
+    private uint _nextMessageId = 1;
+    private Dictionary<uint, ChatRecord> _messages = new();
+    private Dictionary<string, Dictionary<uint, ChatRecord>> _playerMessages = new();
+    private Dictionary<string, string> _userNamesToIds = new();
+
+    public override void Initialize()
+    {
+        Refresh();
+
+        _replay.RecordingFinished += _ =>
+        {
+            // TODO: resolve https://github.com/space-wizards/space-station-14/issues/25485 so we can dump the chat to disc.
+            Refresh();
+        };
+    }
+
+    /// <summary>
+    /// Adds an <see cref="IChatEvent"/> to the repo and raises it with a UID for consumption elsewhere.
+    /// </summary>
+    /// <param name="ev">The event to store and raise</param>
+    /// <returns>If storing and raising succeeded.</returns>
+    public bool Add(IChatEvent ev)
+    {
+        if (!_player.TryGetSessionByEntity(ev.Sender, out var session))
+        {
+            return false;
+        }
+
+        var messageId = _nextMessageId;
+
+        _nextMessageId++;
+
+        ev.Id = messageId;
+
+        var location = new Vector2();
+        var map = "";
+
+        if (TryComp<TransformComponent>(ev.Sender, out var comp))
+        {
+            location = comp.Coordinates.Position;
+
+            if (comp.MapUid != null)
+            {
+                map = Name(comp.MapUid.Value);
+            }
+        }
+
+        var storedEv = new ChatRecord
+        {
+            UserName = session.Name,
+            UserId = session.UserId.UserId.ToString(),
+            EntityName = Name(ev.Sender),
+            Location = location,
+            Map = map,
+            StoredEvent = ev
+        };
+
+        _messages[messageId] = storedEv;
+
+        if (!_playerMessages.TryGetValue(storedEv.UserId, out var set))
+        {
+            set = new Dictionary<uint, ChatRecord>();
+            _playerMessages[storedEv.UserId] = set;
+        }
+
+        set.Add(messageId, storedEv);
+
+        _userNamesToIds[storedEv.UserName] = storedEv.UserId;
+
+        RaiseLocalEvent(ev.Sender, new MessageCreatedEvent(ev), true);
+
```

Comment:
Does this need to be raised as broadcast?

Reply:
If my understanding of the bus is correct, raising as a broadcast is a one-to-many pattern where as raising it without that flag means it's raised and handled in series?

YAGNI so I can remove it for the time being

Reply:
I'll need `broadcast: true` then, as listening for a created message should be done by anything that cares, even if it's not directly related to a component.